Issue37 text style options refactor

ramirezStefano opened the pull request:
All the requirements and specifications of the issue were completed, awaiting review of the current implementation.

ramirezStefano committed:
refactor: changed to flex wrapper, change to toggleButtons, labels rem

re # 37

ramirezStefano committed:
feat: added ColorPicker component to TextStyleOptions component and UI

re #37

ramirezStefano committed:
refactor: reverted to original non optional ColorPicker prop

re #37

ramirezStefano committed:
refactor: changed logic to fit the new components, storybook tested

re #37

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,66 @@
 const TextStyleOptions: FC<TextStyleProps> = ({ isBold, isItalics, onChange }) => {
-    const [selected, setSelected] = React.useState<Selection>(() => {
-        const initialValue: string[] = [];
+    const [selectedBold, setSelectedBold] = React.useState<boolean | undefined>(isBold);
+    const [selectedItalics, setSelectedItalics] = React.useState<boolean | undefined>(isItalics);
```
Quitemos todos los `React.*` usages - en lugar deberiamos estar importando `useState` y `useEffect` de React 

```javascript
import React, { FC, useEffect, useState } from "react";
```

```suggestion
    const [selectedBold, setSelectedBold] = useState<boolean | undefined>(isBold);
    const [selectedItalics, setSelectedItalics] = useState<boolean | undefined>(isItalics);
```

eduardo-vargas replied:
On first use, si los props no están definidos, el `state` es `undefined`, cambiemoslo a que solo acepte `boolean`, [este link debería ser useful](https://stackoverflow.com/questions/31155477/convert-truthy-or-falsy-to-an-explicit-boolean-i-e-to-true-or-false) 
```suggestion
    const [selectedBold, setSelectedBold] = React.useState<boolean>(isBold);
    const [selectedItalics, setSelectedItalics] = React.useState<boolean>(isItalics);
```

<img width="289" alt="image" src="https://user-images.githubusercontent.com/52969178/231620478-d40b7f28-e5bb-439d-9c1e-ad182cc53f3e.png">

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,66 @@
-
-        if (isItalics) {
-            initialValue.push("italics");
+    const handleBoldChange = (value: boolean) => {
+        setSelectedBold(value);
+        isBold = value;
+        if (onChange) {
+            onChange({
+                isBold: value,
+                isItalics: selectedItalics,
+            });
+            console.log(`isBoldPropValue: ${isBold}`);
```
los console.logs tambien se pueden ir

eduardo-vargas replied:
Estas dos funciones son demasiado similares, intentemos brainstorm como hacer que sea solo una función, al rato le podemos meter otro parameter al signature que los distinga 

<img width="1044" alt="image" src="https://user-images.githubusercontent.com/52969178/231620230-8d4133c7-8501-49dc-b331-6b1d6e7ee387.png">

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,66 @@
+                isBold: selectedBold,
+                isItalics: value,
+            });
+            console.log(`isItalicsPropValue: ${isItalics}`);
```
another one

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,66 @@
         }
     };
 
+    React.useEffect(() => {
+        isBold = selectedBold;
+        isItalics = selectedItalics;
+        console.log(`isBold: ${isBold} and isItalics:${isItalics}`);
+    }, [selectedBold, selectedItalics]);
```
🤔  que hace este `useEffect`? 

Los values que se pasan mediante `props` no deberian ser cambiados, los props se pasan downwards pero no deberian ser mutated.

Si los quisieramos cambiar se podria hacer algo como

```javascript 
let copyOfPropValue = propValue;

copyOfPropValue = something else;
```

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,66 @@
-            <ActionGroup
-                isEmphasized
-                density="compact"
+        <Flex direction="row">
+            <ToggleButton isQuiet aria-label="BoldButton" isSelected={selectedBold} onChange={handleBoldChange}>
+                <TagBold />
+            </ToggleButton>
+            <ToggleButton
```
Los botones están muy pegados, agreguemos un gap 

<img width="353" alt="image" src="https://user-images.githubusercontent.com/52969178/231619595-e4c0f4e4-e2ad-42a6-83b3-d388676523ff.png">

eduardo-vargas replied:
También estan en blanco/negro, hagamos que estén [azul cuando están seleccionados](https://react-spectrum.adobe.com/react-spectrum/ToggleButton.html#emphasized)

eduardo-vargas reviewed (changes requested):
Looks good! Some changes

ramirezStefano committed:
refactor: added anonymous function to match expected signature

changed the buttons colors, added spacing and all are isQuiet. Removed useless useEffect and logs

re #37

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,55 @@
+    const handleButtonChange = (valueBold: boolean, valueItalics: boolean) => {
+        setSelectedBold(valueBold);
+        setSelectedItalics(valueItalics);
+        isBold = valueBold;
+        isItalics = valueItalics;
```
Estos dos assignments a los values de los props se pueden (tienen que) ir, los props son solo para initialize el componente pero los valores de esa variable en si no deberian ser cambiados

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,55 @@
-                density="compact"
+                aria-label="BoldButton"
+                isSelected={selectedBold}
+                onChange={selectedBold => handleButtonChange(selectedBold, selectedItalics)}
```
Super use of anonymous function! Nada más que este variable name de `selectedBold` está confusing porque el valor del state hook arriba también es `selectedBold`. 

El value que el onChange nos está pasando es diferente y `handleButtonChange` internamente se encarga de usar el `setSelectedBold` para actualizar el state.

```suggestion
                onChange={newBoldValue => handleButtonChange(newBoldValue, selectedItalics)}
```

eduardo-vargas commented on src/components/TextStyleOptions/TextStyleOptions.tsx:
```diff
@@ -1,62 +1,55 @@
+                isEmphasized
+                aria-label="ItalicsButton"
+                isSelected={selectedItalics}
+                onChange={selectedItalics => handleButtonChange(selectedBold, selectedItalics)}
```
Similar acá 
```suggestion
                onChange={newItalicsValue => handleButtonChange(selectedBold, newItalicsValue)}
```

ramirezStefano committed:
refactor: changed anonymous argument name and in handler paramenter

The assigned prop values in TextStyleOptions have been removed.

re #37
